### PR TITLE
[metabase-lib] Port most of `metabase.util` to CLJC

### PR DESCRIPTION
**Pull request by bshepherdson**
A few JVM-specific functions are now in `metabase.util.jvm`, but they're
re-exported from `metabase.util.`

This is preliminary work for further metabase-lib efforts, which will want many functions from `metabase.util.*`.

### CODE REVIEW

File: .clj-kondo/config.edn [hunk 1]
```diff
@@ -6,7 +6,7 @@
     (clojure.core.logic/fresh)
     (clojure.core.logic/matcha)
     (clojure.core.logic/run)
-    (clojure.test/is [query= sql=])
+    (clojure.test/is [query= re= sql=])
```
Comment (camsaul) [inline .clj-kondo/config.edn:9]: This is in the Kondo config for the Hawk test runner so you don't need to have it here too

File: .clj-kondo/config.edn [hunk 2]
```diff
@@ -625,34 +628,36 @@
...
+   metabase-enterprise.serialization.test-util/with-world                          macros.metabase-enterprise.serialization.test-util/with-world
+   metabase-enterprise.test/with-gtaps                                             macros.metabase-enterprise.sandbox.test-util/with-gtaps
+   metabase.api.card-test/with-ordered-items                                       macros.metabase.api.card-test/with-ordered-items
+   metabase.api.collection-test/with-collection-hierarchy                          macros.metabase.api.collection-test/with-collection-hierarchy
+   metabase.api.collection-test/with-some-children-of-collection                   macros.metabase.api.collection-test/with-some-children-of-collection
+   metabase.api.common/define-routes                                               macros.metabase.api.common/define-routes
+   metabase.api.embed-test/with-embedding-enabled-and-temp-card-referencing        macros.metabase.api.embed-test/with-embedding-enabled-and-temp-card-referencing
+   metabase.api.embed-test/with-embedding-enabled-and-temp-dashcard-referencing    macros.metabase.api.embed-test/with-embedding-enabled-and-temp-dashcard-referencing
```
Comment (camsaul) [inline .clj-kondo/config.edn:643]: Maybe this is nitpicky but I think the old indentation was correct. You have a lot of spaces now

File: test/metabase/actions_test.clj [hunk 3]
```diff
@@ -170,6 +170,13 @@
         (is (thrown-with-msg? Exception #"Value does not match schema:.*"
                               (actions/perform-action! action (dissoc request-body :query))))))))
 
+;; TODO Regex literals can't live under `mt/test-drivers` or any other macro with a clj-kondo hook or `:macroexpand`
+;; rule until https://github.com/clj-kondo/clj-kondo/issues/1987 is fixed and released, so these are defined separately.
+(def ^:private multi-update-regex
+  #"Sorry, this would update [\d,]+ rows, but you can only act on 1")
+(def ^:private multi-delete-regex
+  #"Sorry, this would delete [\d,]+ rows, but you can only act on 1")
```
Comment (camsaul) [inline test/metabase/actions_test.clj:178]: newline between forms please <3

File: test/metabase/query_processor_test/date_bucketing_test.clj [hunk 4]
```diff
@@ -1244,34 +1244,37 @@
 ;; TODO -- is this really date BUCKETING? Does this BELONG HERE?!
 (deftest june-31st-test
   (testing "What happens when you try to add 3 months to March 31st? It should still work (#10072, #21968, #21969)"
-    ;; only testing the SQL drivers for now since I'm not 100% sure how to mock this for everyone else. Maybe one day
-    ;; when we support expressions like `+` for temporal types we can do an `:absolute-datetime` plus
-    ;; `:relative-datetime` expression and do this directly in MBQL.
-    (mt/test-drivers (filter #(isa? driver/hierarchy (driver/the-initialized-driver %) :sql)
+    ;; TODO Regex literals can't live under `mt/test-drivers` or any other macro with a clj-kondo hook or `:macroexpand`
+    ;; rule until https://github.com/clj-kondo/clj-kondo/issues/1987 is fixed and released. So this is bound in a let.
```
Comment (camsaul) [inline test/metabase/query_processor_test/date_bucketing_test.clj:1248]: Why don't we just keep `test-drivers` as a hook if it's such a problem? I think I was running into the same "unused value" issues you were and I fixed it like this in my branch

https://github.com/metabase/metabase/pull/28017/files#diff-42e5d16078af5746c40fe89161de9f1153463de76fcade9070051c0a7aac7cb7R43-R51

File: .clj-kondo/macros/metabase/shared/util/namespaces.clj [hunk 5]
```diff
@@ -1,7 +1,17 @@
...
    `(def ~name "docstring" ~sym)))
+
+(defmacro import-fns
+  "Kondo macro replacement for [[metabase.shared.util.namespaces/import-fns]]."
+  [& pairs]
+  `(vector ~@(for [[from & syms] pairs
+               sym           syms
+               :let [args (gensym)]]
+           (symbol (name from) (name sym))
+           #_`(def ~sym "docstring" (fn [& ~args] (apply ~(symbol (name from) (name sym)) ~args))))))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply .clj-kondo/macros/metabase/shared/util/namespaces.clj:17]: Oh nevermind I'm being dumb I didn't realize this was just the Kondo version of the macroexpansion. It looks like you're not using this config anyway tho?

**Review by camsaul — APPROVED** [review]
I'm 👍 for everything here except for the breaking of regexes inside of `mt/test-drivers`... it seems like the cure is worse than the disease. If we can't make the `:unused-value` linter work I'd rather just disable it until the upstream issues are fixed rather than mangle a bunch of tests

File: .clj-kondo/config.edn (hunk 1 above)
Comment (bshepherdson) [reply .clj-kondo/config.edn:9]: Done.

File: .clj-kondo/config.edn (hunk 2 above)
Comment (bshepherdson) [reply .clj-kondo/config.edn:643]: Fixed.

File: .clj-kondo/macros/metabase/shared/util/namespaces.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (bshepherdson) [reply .clj-kondo/macros/metabase/shared/util/namespaces.clj:17]: Yeah, it's dead code now. Removed.

File: test/metabase/actions_test.clj (hunk 3 above)
Comment (bshepherdson) [reply test/metabase/actions_test.clj:178]: These are reverted.

File: test/metabase/actions_test.clj [hunk 6]
```diff
@@ -407,66 +424,72 @@
...
+                                                               :arg
+                                                               rows}))]
+            (testing "Initial values"
+              (is (= [[1 "African"]
+                      [2 "American"]
+                      [3 "Artisan"]]
+                     (first-three-categories))))
+            (testing "Should report the index of input rows with errors in the data warehouse"
+              (let [error-message-regex (case driver/*driver*
+                                          :h2       #"^NULL not allowed for column \"NAME\""
```
[the remark this answers is not in the document]
Comment (bshepherdson) [reply test/metabase/actions_test.clj:453]: For the record it was only those with backslashes in. This is reverted anyway.

File: test/metabase/query_processor_test/date_bucketing_test.clj (hunk 4 above)
Comment (bshepherdson) [reply test/metabase/query_processor_test/date_bucketing_test.clj:1248]: Done as you suggested.

**Pull request merged**